UNIXFS.md

```diff
@@ -59,6 +64,27 @@ For files that are comprised of more than a single block, the 'Type' field will
 
 This data is serialized and placed inside the 'Data' field of the outer merkledag protobuf, which also contains the actual links to the child nodes of this object.
 
+## Metadata
+
+There are three ways to specify file metadata:
+
+1. The repeated `metadata` field in a directory applies metadata to each file in the directory.
+2. An intermediary node with a `Type` of `Metadata` applies metadata to an individual file. While this feature should be supported, it has been deprecated.
+3. An optional `defaultMetadata` field to specify the _default_ metadata for files in the directory. If unspecified, the default mode is 0755 and the default modification time is the epoch.
```

Comment:
This can save us a _ton_ of space as most files in a directory have the same mode. However, it _is_ slightly complicated.

@mib-kd743naq suggested keeping a reverse mapping of, e.g., `mode -> [file offsets]` but, while very compact, that felt too complicated.

---

On the other hand, we could _also_ just rely on future compression. Repeated identical metadata fields will compress _really_ well.

Reply:
I'll channel @mib-kd743naq's alter-ego here ;P The idea of `defaultMetadata` looks deceptively simple, but would be hard to make "convergent". The typical mode of producing UFS1.5 datasets would be tar-file ( or equivalent ) importers. A tar importer works best ( and most faithfully ) in linear streaming fashion, which in turn means that the "best" defaultMetadata is difficult to know ahead of time. This is especially important for mega-directories like  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm/boot/dts and is even harder to determine in cases of sharded directories.

The flip side of doing "deduplication of metadata" as we go is to simply keep a map during encoding of "what metadata have we seen so far" keyed off the encoded protobuf content itself, together with a plain integer increment counter.

With all that said - it **appears to NOT matter**. Using the dataset of the current kernel, with each individual entry decorated with the proper metadata based on a spec similar[*] to the proposed the numbers are **way** closer than I anticipated. Also, yes, transparent block compression will help a lot here, but there still will be a lot of indirection for a compressor to deal with, and there is nothing simpler than simply having nothing extra included ;) Additionally transparent compression is a separate :canofworms: that may take longer than UFS2.0 to shake out :(

So the data based on https://cdn.kernel.org/pub/linux/kernel/v5.x/linux-5.3.7.tar.sign

To see individual directory makeups simply run:
```
for h in \
  bafybeif7ceybwlonoct3bnjtnnfookm2iezslo2ypwrsae6ilkc2ucvsfm \
  bafybeigoe7cnxqy6hjlcci3ae62l5fmws5pxeq62d77lhbvgphhm5ugtnq \
  bafybeico6ybaiaqohfsydgqxexwcjb4vnlhzojyxhjcw6s3xhm7weaikb4 \
  bafybeibj2hpul5vgomel56rl6xf63rviyym4go5pg2fmwwqioaltshh6fa
do  tgt=$h/linux-5.3.7/Documentation/PCI/endpoint
  echo -e "\n\nhttps://ipfs.io/ipfs/$tgt\n==============\n"
  ipfs block get $( ipfs resolve $tgt | cut -b 7- ) | protoc --decode_raw 
done | less
```


And the stats:

```
dumb ( each directory entry gets an exactly one "struct 7" ):

ipid                    : bafybeif7ceybwlonoct3bnjtnnfookm2iezslo2ypwrsae6ilkc2ucvsfm
count_data_blocks       :         230,613
count_link_blocks       :          42,341
count_all_blocks        :         272,954
size_data_blocks        :     843,884,873
size_link_blocks        :      13,694,638
size_all_blocks         :     857,579,511
size_payload            :       ^ dedup ^



stebalien-first ( Stebalien's  proposal, if more than 2 entries are present in a directory, the first one becomes the "default struct 8" ):

ipid                    : bafybeigoe7cnxqy6hjlcci3ae62l5fmws5pxeq62d77lhbvgphhm5ugtnq
count_data_blocks       :         230,613
count_link_blocks       :          42,341
count_all_blocks        :         272,954
size_data_blocks        :     843,884,873
size_link_blocks        :      13,036,969
size_all_blocks         :     856,921,842
size_payload            :       ^ dedup ^



stebalien-most ( Steven's proposal, if more than 2 entries are present in a directory, an analysis takes place of which is the most-encountered, and that one becomes default ):

ipid                    : bafybeico6ybaiaqohfsydgqxexwcjb4vnlhzojyxhjcw6s3xhm7weaikb4
count_data_blocks       :         230,613
count_link_blocks       :          42,341
count_all_blocks        :         272,954
size_data_blocks        :     843,884,873
size_link_blocks        :      13,013,814
size_all_blocks         :     856,898,687
size_payload            :       ^ dedup ^



olcbean ( the proposal to just have a running reverse index per metadata blob ):

ipid                    : bafybeibj2hpul5vgomel56rl6xf63rviyym4go5pg2fmwwqioaltshh6fa
count_data_blocks       :         230,613
count_link_blocks       :          42,341
count_all_blocks        :         272,954
size_data_blocks        :     843,884,873
size_link_blocks        :      12,941,959
size_all_blocks         :     856,826,832
size_payload            :       ^ dedup ^
```

* Once I am back from a trip we really need to chat some time next week why I keep pushing for a slightly different organization of the metadata.

Reply:
Note: this is unixfsv1.5. We don't have to be _perfect_, just good enough.

Given this data, I'm starting to think that we might just want to _drop_ the default metadata field and default to 0644 for files and 0755 for directories. Those numbers show a 0.09% overhead which I can easily live with.

> in linear streaming fashion

Note: the reverse index is just as tricky to compute.

Reply:
> This is mostly because of huge/sharded directories right? Otherwise, seems reasonably straightforward.

I'm not concerned about sharded directories. We'd just store the reverse map in every terminal shard.

My main concern is that I'd _like_ to just stick this metadata _inside_ the directory entries in unixfsv2, right next to the links to the files.

1. This is "cleaner" (subjectively).
2. It makes it easy to move a directory entry from one directory to another (just copy the node).
3. When decoding the directory, we'd need to reverse the reverse index (turn it back into something where we can easily lookup metadata by file name and/or offset). At the moment, in all of the proposals for UnixFSv2 that I've seen, we'll be able to define the mapping between go structs and encoded IPLD objects using a nice IPLD schema. If we use this reverse map, we're going to either:
   1. Need to introduce a second step and a second intermediate object.
   2. Turn a "reverse index" into an advanced layout.

Reply:
One data point from Peergos, though I'm not sure how relevant it is to unixfsv2 given the lack of encryption: we decided to put file metadata (name, modification time, mime type etc.) on the file itself, rather than the directory. This is because of using capability based access control - we want to be able to easily grant someone access to modify a file without them being able to read or modify or even identify sibling files in the same directory. So our directories are just encrypted links to files (and the encrypted name of the directory) which might be under a different IPNS signing key and an encrypted link to any subsequent block in the same directory (large directories).

Reply:
That's good to know.

In that case, we'd probably have a special "encrypted" file type. Normal files will have a "data" field referring to the actual file data but _this_ file type would have a "file" field pointing to the actual encrypted file.

However, we'd still likely want _some_ unencrypted metadata ("file type", capabilities?, etc.).

Reply:
We take it even further. Unless you have a read cap to a directory we don't want you to even be able to know how many files are in the directory (we pad and bulk encrypt all the child links together), and we even go so far as to make a directory indistinguishable from a small file at the ipld level (without a read cap). That way you can grant read access to a file, but even with that capability the only thing you can figure out about the parent directory is the name (to get a well defined path).

Reply:
Ah, interesting. so you encrypt the directory entries themselves. At that point, I think we'll need to change the structures entirely.

Reply:
unixfsv3 then :-)